altos/ao-tools/ao-telem/ao-telem.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-TELEM 1 "ao-telem" ""
.SH NAME
ao-telem \- Analyze a telemetry log
.SH SYNOPSIS
.B "ao-telem"
[\--crc]
{flight.telem} ...
.SH OPTIONS
.TP
\-c | --crc
This option makes ao-telem analyze records with CRC errors. By
default, these records are skipped.
.SH DESCRIPTION
.I ao-telem
reads the specified telemetry log and display the contents of each
record.
.SH AUTHOR
Keith Packard


altos/ChangeLog
commit dbcb68f684a96d13efeb9993685f61b27b45e7dc
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Aug 12 00:21:43 2017 -0400

    update docinfo file for 1.8 release

commit dfb9568866ba250a02039cc3429d5ca0603a60d9
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:10:23 2017 -0400

    Note that telebt 4.0 files need to go to altusmetrumllc dir
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 6b083e90d5b8de215f2f303f18b5c4ac9185622d
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:06:48 2017 -0400

    Bump version to 1.8
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 243c1e519d41b45765d4edf5f608655d6ab01eff
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:06:26 2017 -0400

    doc: Add 1.8 release notes
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7b67dcdd90317845261e51a1adb2404dcb0f2507
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:05:50 2017 -0400

    altosui, telegps: Add TeleBT v4.0 firmware to app packages
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 5fdce5b4a70ae59f8ac1ca10d9fadbbbd62412f3
Merge: f1bbf1bd 2a3db289
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:45:25 2017 -0400

    Merge remote-tracking branch 'origin/master'

commit f1bbf1bd18d52a084af714a022b12bfc3c76374c
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:40:41 2017 -0400

    ao-bringup: Allow dfu_util to "fail" while flashing chaoskey
    
    Sometimes it isn't happy about the exit status, but if the device
    comes up and works OK, we don't care.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit fbe87bd18308cab62eda6d5e956a8618095ed33b
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:39:13 2017 -0400

    altoslib: Set time in state for KML output correctly
    
    Use new computed state_time value.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c7b34e6691b632199bc7802c2e4dea4de992578e
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:49:26 2017 -0700

    libaltos: Test both bluetooth and USB APIs with cjnitest
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 52bbfa9cd46ca27a4de937e89d48246a4e94c8a6
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:47:40 2017 -0700

    altos/test: Fix AES key format in test code
    
    Use temp variable instead of stepping on the AES name.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9eaf917707287af57c2dadfc82d6d415b0964329
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:46:24 2017 -0700

    altos/test: Fix printf format for int64 arithmetic test
    
    Use 'long long' and %lld for 64-bit values when printing.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 198a3d4d5f51de5c0f0413299582479cde4e177f
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:37:27 2017 -0700

    altosui: Rename AltosConfig -> AltosConfigFC
    
    Now the class name says that this configures the flight computer.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c7c2fc2d85414fefda0a7948a6c4e38f65140861
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:38:09 2017 -0700

    altosui: Adapt to flight stats time value changes
    
    There aren't state time values anymore as those don't work when you
    have multiple motors. Instead, 'boost_time' is when the rocket left
    the pad and 'landed_time' is when it touched down. Use these new
    values in the --summary output.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2a3db28953f5cd981c519a917cdcd9f48dcb08e9
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:49:21 2017 -0600

    add TeleBT v4.0 to the list of stuff we build

commit 04560a7cd6a64e742954a2d0fbd7563e2c310402
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:46:55 2017 -0600

    ignore more build products in telebt-v4.0 tree

commit c8dbcaf69cd538a31ab6e2b568237ae7c8656a9a
Merge: 0cbfa444 ea3b5815
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:44:03 2017 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 0cbfa444a9f9159cb509bb47ca5590fc1d709f64
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:43:38 2017 -0600

    clean up telefireone name references

commit 3a1512b605a81ebee23cc6b7980b709a54cb43c9
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:42:54 2017 -0600

    rename existing turnon_telebt script, create a new one for v4.0

commit ea3b5815b27005b2f4c3034715f656d28ea8534e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:16:55 2017 -0700

    libaltos: Add btle test program
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 545bdc1f8046b08e237307d97d0044ddae9095b8
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:14:31 2017 -0700

    altos/test: Add EasyMini test mode
    
    Use baro-only mode, parse easymini logs.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2969f0af52dfbdea7593b208e510c7b17227064c
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:09:55 2017 -0700

    altoslib: fix adjustment of landed time
    
    Push landed time out if we computed silliness.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 30b308ad0189b4aa9dd5cc7ee876b55fee355a56
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:05:45 2017 -0700

    Set version to 1.7.1 for TeleBT firmware build
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b923e25a8a49dda40e62070dd2df6ff9130d4b38
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Jun 27 23:39:16 2017 -0700

    altos/telebt-v4.0: Use HW flow control to BT module
    
    Now that the stmf0 HW flow control works...
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e64438271e481ca9f9eeb0ce2b1f9eca071b7ad4
Author: Keith Packard <keithp@keithp.com>

altos/ao-tools/ao-telem/ao-telem.c
/*
 * Copyright © 2011 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#define _GNU_SOURCE
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include "cc.h"

static const struct option options[] = {
	{ .name = "crc", .has_arg = 0, .val = 'c' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--crc] {flight.telem} ...\n",program);
	exit(1);
}

#define bool(b)	((b) ? "true" : "false")

static int ignore_crc;

static void
process(FILE *file)
{
	char	line[1024];
	int c;

	while (fgets(line, sizeof (line), file)) {
		union ao_telemetry_all telem;
		char call[AO_MAX_CALLSIGN+1];
		char version[AO_MAX_VERSION+1];

		if (cc_telemetry_parse(line, &telem)) {
			int rssi = (int8_t) telem.generic.rssi / 2 - 74;

			printf ("serial %5d rssi %d status %02x tick %5d type %3d ",
				telem.generic.serial, rssi, telem.generic.status,
				telem.generic.tick, telem.generic.type);
			if (!ignore_crc && (telem.generic.status & (1 << 7)) == 0) {
				printf ("CRC error\n");
				continue;
			}
			switch (telem.generic.type) {
			case AO_TELEMETRY_SENSOR_TELEMETRUM:
			case AO_TELEMETRY_SENSOR_TELEMINI:
			case AO_TELEMETRY_SENSOR_TELENANO:
				printf ("state %1d accel %5d pres %5d ",
					telem.sensor.state, telem.sensor.accel, telem.sensor.pres);
				printf ("accel %6.2f speed %6.2f height %5d ",
					telem.sensor.acceleration / 16.0,
					telem.sensor.speed / 16.0,
					telem.sensor.height);
				printf ("ground_pres %5d ground_accel %5d accel_plus %5d accel_minus %5d\n",
					telem.sensor.ground_pres,
					telem.sensor.ground_accel,
					telem.sensor.accel_plus_g,
					telem.sensor.accel_minus_g);
				break;
			case AO_TELEMETRY_CONFIGURATION:
				memcpy(call, telem.configuration.callsign, AO_MAX_CALLSIGN);
				memcpy(version, telem.configuration.version, AO_MAX_VERSION);
				call[AO_MAX_CALLSIGN] = '\0';
				version[AO_MAX_CALLSIGN] = '\0';
				printf ("device %3d flight %5d config %3d.%03d delay %2d main %4d log_max %5d",
					telem.configuration.device,
					telem.configuration.flight,
					telem.configuration.config_major,
					telem.configuration.config_minor,
					telem.configuration.apogee_delay,
					telem.configuration.main_deploy,
					telem.configuration.flight_log_max);
				printf (" call %8s version %8s\n", call, version);
				break;
			case AO_TELEMETRY_LOCATION:
				printf ("sats %d flags %s%s%s%s",
					telem.location.flags & 0xf,
					(telem.location.flags & (1 << 4)) ? "valid" : "invalid",
					(telem.location.flags & (1 << 5)) ? ",running" : "",
					(telem.location.flags & (1 << 6)) ? ",date" : "",
					(telem.location.flags & (1 << 7)) ? ",course" : "");
				printf (" alt %5d lat %12.7f lon %12.7f",
					AO_TELEMETRY_LOCATION_ALTITUDE(&telem.location),
					telem.location.latitude / 1e7,
					telem.location.longitude / 1e7);
				if ((telem.location.flags & (1 << 6)) != 0) {
					printf (" year %2d month %2d day %2d",
						telem.location.year,
						telem.location.month,
						telem.location.day);
					printf (" hour %2d minute %2d second %2d",
						telem.location.hour,
						telem.location.minute,
						telem.location.second);
				}
				printf (" pdop %3.1f hdop %3.1f vdop %3.1f mode %d",
					telem.location.pdop / 5.0,
					telem.location.hdop / 5.0,
					telem.location.vdop / 5.0,
					telem.location.mode);
				if ((telem.location.flags & (1 << 7)) != 0)
					printf (" ground_speed %6.2f climb_rate %6.2f course %d",
						telem.location.ground_speed / 100.0,
						telem.location.climb_rate / 100.0,
						telem.location.course * 2);
				printf ("\n");
				break;
			case AO_TELEMETRY_SATELLITE:
				printf ("sats %d", telem.satellite.channels);
				for (c = 0; c < 12 && c < telem.satellite.channels; c++) {
					printf (" sat %d svid %d c_n_1 %d",
						c,
						telem.satellite.sats[c].svid,
						telem.satellite.sats[c].c_n_1);
				}
				printf ("\n");
				break;
			case AO_TELEMETRY_MEGA_SENSOR_MPU:
			case AO_TELEMETRY_MEGA_SENSOR_BMX160:
				printf ("orient %3d accel %5d pres %9d temp %5d accel_x %5d accel_y %5d accel_z %5d gyro_x %5d gyro_y %5d gyro_z %5d mag_x %5d mag_y %5d mag_z %5d\n",
					telem.mega_sensor.orient,
					telem.mega_sensor.accel,
					telem.mega_sensor.pres,
					telem.mega_sensor.temp,
					telem.mega_sensor.accel_x,
					telem.mega_sensor.accel_y,
					telem.mega_sensor.accel_z,
					telem.mega_sensor.gyro_x,
					telem.mega_sensor.gyro_y,
					telem.mega_sensor.gyro_z,
					telem.mega_sensor.mag_x,
					telem.mega_sensor.mag_y,
					telem.mega_sensor.mag_z);
				break;
			case AO_TELEMETRY_COMPANION:
				printf("board_id %3d update_period %3d channels %2d",
				       telem.companion.board_id,
				       telem.companion.update_period,
				       telem.companion.channels);
				for (c = 0; c < telem.companion.channels; c++)
					printf(" %6d", telem.companion.companion_data[c]);
				printf("\n");
				break;
			case AO_TELEMETRY_MEGA_DATA:
				printf ("state %1d v_batt %5d v_pyro %5d ",
					telem.mega_data.state,
					telem.mega_data.v_batt,
					telem.mega_data.v_pyro);
				for (c = 0; c < 6; c++)
					printf ("s%1d %5d ", c,
						telem.mega_data.sense[c] |
						(telem.mega_data.sense[c] << 8));

				printf ("ground_pres %5d ground_accel %5d accel_plus %5d accel_minus %5d ",
					telem.mega_data.ground_pres,
					telem.mega_data.ground_accel,
					telem.mega_data.accel_plus_g,
					telem.mega_data.accel_minus_g);

				printf ("accel %6.2f speed %6.2f height %5d\n",
					telem.mega_data.acceleration / 16.0,
					telem.mega_data.speed / 16.0,
					telem.mega_data.height);

				break;
			case AO_TELEMETRY_METRUM_SENSOR:
				printf ("state %1d accel %5d pres %9d temp %6.2f acceleration %6.2f speed %6.2f height %5d v_batt %5d sense_a %5d sense_m %5d\n",
					telem.metrum_sensor.state,
					telem.metrum_sensor.accel,
					telem.metrum_sensor.pres,
					telem.metrum_sensor.temp / 100.0,
					telem.metrum_sensor.acceleration / 16.0,
					telem.metrum_sensor.speed / 16.0,
					telem.metrum_sensor.height,
					telem.metrum_sensor.v_batt,
					telem.metrum_sensor.sense_a,
					telem.metrum_sensor.sense_m);
				break;
			case AO_TELEMETRY_METRUM_DATA:
				printf ("ground_pres %9d ground_accel %5d accel_plus %5d accel_minus %5d\n",
					telem.metrum_data.ground_pres,
					telem.metrum_data.ground_accel,
					telem.metrum_data.accel_plus_g,
					telem.metrum_data.accel_minus_g);
				break;
			case AO_TELEMETRY_MINI:
				printf ("state %1d v_batt %5d sense_a %5d sense_m %5d pres %9d temp %6.2f acceleration %6.2f speed %6.2f height %5d ground_pres %9d\n",
					telem.mini.state,
					telem.mini.v_batt,
					telem.mini.sense_a,
					telem.mini.sense_m,
					telem.mini.pres,
					telem.mini.temp / 100.0,
					telem.mini.acceleration / 16.0,
					telem.mini.speed / 16.0,
					telem.mini.height,
					telem.mini.ground_pres);
				break;
			default:
				printf("\n");
			}
		}
	}
}

int
main (int argc, char **argv)
{
	int c, i, ret = 0;
	FILE *file;
	while ((c = getopt_long(argc, argv, "c", options, NULL)) != -1) {
		switch (c) {
		case 'c':
			ignore_crc = 1;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	if (optind < argc) {
		for (i = optind; i < argc; i++) {
			file = fopen(argv[i], "r");
			if (!file) {
				perror(argv[i]);
				ret++;
				continue;
			}
			process(file);
			fclose (file);

		}
	} else
		process(stdin);
	return ret;
}
